_posts/2018-10-11-migrating-from-galen-to-backstop.md

```diff
@@ -0,0 +1,257 @@
+
+## Switch
+
+For some time we were running both (Galen and Backstop based) visual regression tests in parallel. After having the Backstop one stabilized, we turned the other one off.
+
+## Git repository and storing reference files
+
+One of the things we perceive as crucial is the possibility to clone the repo of our application and with just a few commands run it or execute the tests. It should be self contained.
+
+What about storing reference files inside the repository, then? Should we keep them there? We decided **not to**, and to put them on the server, where all of the tests are executed. Why? For two key reasons.
+
+### 1. Differences in rendering
+
+Despite the fact, that all of the tests are run in the same browser everywhere (headless Chrome), it is not guaranteed that the rendering will look 100% the same on various OSs. E.g. I am using a macOS, my colleague is using some Linux distro. I bet there will be differences. If he were to submit new reference files, tests on my end would fail. So could the ones executed on the pipeline server (yet another OS).
+
+Therefore we decided to keep the files on the pipeline server, where the whole environment is constant.
+
+Sidenote: we still can run the tests locally. For that it is required to:
+
+1. Make the reference files from `master` branch.
+1. Apply whatever changes in the code.
+1. Run the test(s), and the images created will be checked against the ones created on the same OS.
+
+### 2. Large amount of files
+
+Right now, our reference files weight ~85MB. Let's say something gets changed in the header of the application. Most of the images would have to be refreshed and, say, ~60MB of new ones would be created. What would that mean? Downloading that ~60MB of data on next `git pull`, or whooping ~145MB of data on `git clone`. What if more changes are applied, or reverted? More and more MBs to download. That would be just too much.
+
+Sure, there are ways of bypassing this. One of the solutions is to use [LFS](https://git-lfs.github.com/). Yet, using that extension doesn't fix the no. 1 problem (rendering). We could go a bit further and use [Docker](https://www.docker.com/) to ensure we have the very same environment everywhere, but that just adds too much complexity 'only' for running visual regression tests.
+
+## Final words, learnings, possible improvements
+
+Occasionally we had, and still have, some tests failing from time to time (lazy loaded images, very rarely), but we are satisfied now. The most important thing is that we can ship more often. **We went down from ~30min to less than 5 with our pipeline!**
```

Comment:
Is the migration really the reason for this huge performance increase? Because at the same time a significant amount of tests have been just deleted, which would be also an good explanation for this result.

Reply:
We migrated some of the tests. That is true. I did not mention that. I have to. Thanks!

Reply:
I will try to find out how long it took the CI to pass without the deleted ones ... but I am afraid it (report) will be long gone.

Reply:
We should be able to add those tests if we really need them without loosing much performance. Those tests won't increase the executing time more than 2 minutes in my opinion.

Reply:
I have found that. Before removing some tests it was ~29min for the whole pipeline. After removing them it was ~20min. I guess I can use that number instead. And also, what impact that had:
1. randomly failing tests meant rerunning them again
2. rebasing meant running them again (that is the same for both runners), but there was a high chance for tests to fail (see no. 1)
I guess I will add that as well.